### ISSUE TITLE
Recursive function calls: Stack Overflow, XQDY0054

### ISSUE THREAD

**Issue by ChristianGruen** [BaseXdb/basex#2324]
Based on #1095: The following queries still result in StackOverflow errors:

```xquery
declare variable $f := map {
  'A': function($r, $f) { $r/x ! $f?B(., $f) },
  'B': function($r, $f) { $r/x ! $f?A(., $f) }
};
$f?A(<a/>, $f)
```

```xquery
let $even := fn($n, $self, $odd) {
  $n = 0 and $odd($n - 1, $odd, $self)
}
let $odd := fn($n, $self, $even) {
  $n != 0 or $even($n - 1, $even, $self)  
}
return $even(1, $even, $odd)
```

**Comment by ChristianGruen** [comment]
And another one that raises `XQDY0054` (see https://www.mail-archive.com/basex-talk@mailman.uni-konstanz.de/msg15865.html):

```xquery
declare variable $a := function-lookup(xs:QName('local:x'), 0);
declare variable $b := $a;
declare function local:f() { $b };
local:f()
```

**Comment by ChristianGruen** [comment]
The queries in the first comment were fixed by dcf52450ec6166db340800895bbf940427d75e8f.
